frontends/pytorch/python/torch_mlir_utils/codegen/torch_signature_ods_gen.py

```diff
@@ -0,0 +1,472 @@
+    variant (they may not but we generate patterns to match anyway).
+    """
+    reg_record = self._get_reg_record(reg_name, f"{reg_name}.Tensor")
+    ods_ins, arg_type_flags = self._map_sigtypes(
+        reg_record["arguments"],
+        type_transforms={
+            "Tensor:0": "AnyTorchImmutableTensor",
+        },
+        flag_transforms={
+            ":0": ["kImmutableTensor"],
+        })
+    ods_outs, return_type_flags = self._map_sigtypes(
+        reg_record["returns"],
+        type_transforms={
+            "Tensor:0": "AnyTorchImmutableTensor",
+        },
+        flag_transforms={
+            ":0": ["kImmutableTensor"],
+        })
+    self.ods_emitter.emit_opdef(ods_name,
+                                op_name,
+                                reg_record,
+                                ods_ins=ods_ins,
+                                ods_outs=ods_outs,
+                                traits=["NoSideEffect"])
+    self.impl_emitter.emit_kernel_methods(ods_name,
+                                          reg_record,
+                                          arg_type_flags=arg_type_flags,
+                                          return_type_flags=return_type_flags,
+                                          promote_trailing_out_tensor=True)
+
+  def _get_reg_record(self, *reg_names):
+    """Gets the op-dict for a given registered op name.
+
+    Args:
+      reg_names: The keys in the registry to probe.
+        Typically {kernel_name}.{variant} or {kernel_name} (for empty variant).
+    Returns:
+      Dict of the registration record.
+    """
+    for reg_name in reg_names:
+      record = self.reg_ops.get(reg_name)
+      if record:
+        return record
+    raise ValueError(f"Could not find registry op matching one of {reg_names}")
+
+  def _map_sigtypes(self, siglist: List[Dict], type_transforms: Dict[str, str],
+                    flag_transforms: Dict[str, List[str]]) -> List[Tuple[str]]:
+    """Maps a list of signature entries to ods dags and flag lists.
+
+    The torch signature list contains dicts that minimally have keys 'name' and
+    'type', representing torch names and types. Returns a corresponding
+    list of 2-tuples of (ods_name, ods_type).
+
+    The provided type_transforms is a dict of type substitutions, one of which
+    must match for each entry in the list. The keys can be either a verbatim
+    torch type (i.e. "Tensor") an index in the list (i.e. ":0") or a type and
+    index (i.e. "Tensor:0").
+
+    Similarly, flag_transforms matches its keys in the same way and maps to
+    a list of KernelValueConversion constants that make up arg/return specific
+    conversion flags.
+
+    Returns:
+      - An ods dag list of (ods_name, ods_type) tuples
+      - List of (torch_type, [conversion_flag]) for specifying conversions.
+    """
+    # Generate to ods dag list.
+    ods_dag_list = []
+    for i, sigitem in enumerate(siglist):
```

Comment:
Any chance the PyTorch signature list format is described somewhere you can link to? At this point in the code (and callers of this function), without specific knowledge of that format, it gets really hard to follow.

Reply:
It's a bit endemic to how it operates and is just (for what we care about) kernel name + type list.

What I did is remove the fragile dependence on the (kernel_name, variant), which is not well-defined (the variant is sometimes arbitrary) and reworked the registration keys to be based on the stringified actual signature, same as for PT internally (i.e. 'aten::add(Tensor,Tensor,Scalar)' vs 'aten::add.Tensor'). This is more correct and likely more clear. I also added a did-you-mean lookup for the error message when one doesn't match, because looking in the list is the most common thing I do.

I think this makes the concept more clear without a lot of documentation. Lmk.